mongodbatlas/resource_mongodbatlas_advanced_cluster.go

```diff
@@ -511,6 +515,64 @@ func resourceMongoDBAtlasAdvancedClusterRead(ctx context.Context, d *schema.Reso
 	return nil
 }
 
+func resourceMongoDBAtlasAdvancedClusterUpdateOrUpgrade(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	upgradeRequest := getUpgradeRequest(d)
+
+	if upgradeRequest != nil {
+		upgradeCtx := context.WithValue(ctx, upgradeRequestCtxKey, upgradeRequest)
+		return resourceMongoDBAtlasAdvancedClusterUpgrade(upgradeCtx, d, meta)
+	}
+
+	return resourceMongoDBAtlasAdvancedClusterUpdate(ctx, d, meta)
+}
+
+func resourceMongoDBAtlasAdvancedClusterUpgrade(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*MongoDBClient).Atlas
+	ids := decodeStateID(d.Id())
+	projectID := ids["project_id"]
+	clusterName := ids["cluster_name"]
+
+	var upgradeResponse *matlas.Cluster
+	var err error
+	upgradeRequest := ctx.Value(upgradeRequestCtxKey).(*matlas.Cluster)
+
+	if upgradeRequest == nil {
+		return diag.FromErr(fmt.Errorf("upgrade called without %s in ctx", string(upgradeRequestCtxKey)))
+	}
+
+	err = resource.RetryContext(ctx, 3*time.Hour, func() *resource.RetryError {
+		upgradeResponse, _, err = upgradeCluster(ctx, conn, upgradeRequest, projectID, clusterName)
+		if err != nil {
+			var target *matlas.ErrorResponse
+			if errors.As(err, &target) && target.ErrorCode == "CANNOT_UPDATE_PAUSED_CLUSTER" {
+				clusterRequest := &matlas.AdvancedCluster{
+					Paused: pointy.Bool(false),
+				}
+				_, _, err := updateAdvancedCluster(ctx, conn, clusterRequest, projectID, clusterName)
+				if err != nil {
+					return resource.NonRetryableError(fmt.Errorf(errorClusterAdvancedUpdate, clusterName, err))
+				}
+			}
+			if errors.As(err, &target) && target.HTTPCode == 400 {
+				return resource.NonRetryableError(fmt.Errorf(errorClusterAdvancedUpdate, clusterName, err))
+			}
+		}
+		return nil
+	})
```

Comment:
[q] Why do we upgrade a paused cluster? if the cluster is paused, we may not want to upgrade it since the customer will pay for it

if we really want to upgrade the cluster, we should pause the cluster after the upgrade

Reply:
Sounds like we want to upgrade the cluster, then pause it, thanks for catching that!